android/src/main/kotlin/io/lakscastro/sharedstorage/storageaccessframework/lib/DocumentCommon.kt

```diff
@@ -161,12 +163,24 @@ fun traverseDirectoryEntries(
   rootOnly: Boolean,
   block: (data: Map<String, Any>, isLast: Boolean) -> Unit
 ): Boolean {
-  val childrenUri = DocumentsContract.buildChildDocumentsUriUsingTree(
-    rootUri,
-    DocumentsContract.getTreeDocumentId(rootUri)
-  )
 
-  /// Keep track of our directory hierarchy
+  var childrenUritmp: Uri?;
+
+  // https://stackoverflow.com/questions/41096332/issues-traversing-through-directory-hierarchy-with-android-storage-access-framew
+  // Credit to user: Foobnix
+  // If we were to always use getTreeDocumentId, it would apparently always only list the top level folder even if you request a subfolder
+  try {
+      // for childs and sub child dirs
+       childrenUritmp = DocumentsContract.buildChildDocumentsUriUsingTree(rootUri, DocumentsContract.getDocumentId(rootUri));
+  } catch (e: Exception) {
+      // for parent dir
+       childrenUritmp = DocumentsContract.buildChildDocumentsUriUsingTree(rootUri, DocumentsContract.getTreeDocumentId(rootUri));
+  }
```

Comment:
This makes a lot of sense. I knew something was missing in this function since the subfolders wasn't allowed to be listed even if we had permisions over a parent folder.

Most of times testing and debugging this kind of operation is not trivial and requires a lot of time, thanks for the efforts! I'll make sure to publish into the next release and since it's not a breaking change and fixes a current behavior.